### PR TITLE
Handle mul wikisource pages

### PR DESCRIPTION
**Pull request by Parthiv-M**
Modify Wikidata query to include
mul wikisource and handle it in
related methods as well

Bug: [T345325](https://phabricator.wikimedia.org/T345325)

### CODE REVIEW

**Comment by samwilson** [comment]
With this, `WikisourceApi::fetchWikisources()` returns a lot of duplicates for `mul`, e.g. the output of `examples/list.php` includes the following:

```
* pa - ਪੰਜਾਬੀ - Q28861316
* ml - മലയാളം - Q22004659
* li - Limburgs - Q24577674
* mul - لغات متعددة - Q18198097
* mul - distintes llingües - Q18198097
* mul - çoxdilli - Q18198097
* mul - бер нисә тел - Q18198097
* mul - некалькі моў - Q18198097
* mul - некалькі моваў - Q18198097
```

It's slightly odd, because this library never really thought about Multilingual Wikisource to start with (sorry! that's my fault), and so it has a language code and name for each Wikisource (e.g. `Wikisource::$langName`). For Multilingual Wikisource these should probably return `mul` and a localised name, perhaps from  *multiple languages* (Q20923490). That's not particularly satisfactory, but do you think it'll make sense enough? Either that, or we could explicitly separate mulwikisource from the others, so consuming code would have to change.

**Comment by Parthiv-M** [comment]
Ah, I did see this in the output, but was not sure what to make of it. 

I'm not entirely sure I get what you mean by _localised name_. Does that mean we don't need to show the names `некалькі моў`, `некалькі моваў`, etc.?

How would the output be formatted then?

File: src/WikisourceApi.php [hunk 1]
```diff
@@ -104,15 +104,24 @@ public function fetchWikisources( $cacheLifetime = null ) {
...
+				. "?lang rdfs:label ?langName . FILTER(LANG(?langName) = ?langCode) ."
+				// perform a union of the results
+				. " } UNION { "
+				// Instance of Wikisource language edition.
+				. "?item wdt:P31 wd:Q15156455 ."
+				// Wikimedia language code.
+				. "?item wdt:P424 ?langCode ."
+				// language of work or name is multiple languages
+				. "?item wdt:P407 wd:Q20923490 . } }";
+			echo $query;
```
Comment (samwilson) [inline src/WikisourceApi.php]: Remove debugging statement.

File: src/WikisourceApi.php [hunk 2]
```diff
@@ -126,6 +135,9 @@ public function fetchWikisources( $cacheLifetime = null ) {
 			$ws = new Wikisource( $this, $this->logger );
 			$ws->setLanguageCode( $langInfo['langCode'] );
 			$ws->setLanguageName( $langInfo['langName'] );
```
Comment (samwilson) [inline src/WikisourceApi.php]: This line is throwing a warning for the `mul` row.

> Warning: Undefined array key "langName" in wikisource-api/src/WikisourceApi.php on line 137

File: src/WikisourceApi.php [hunk 3]
```diff
@@ -126,6 +135,9 @@ public function fetchWikisources( $cacheLifetime = null ) {
...
+			if ( $langInfo['langCode'] == "mul" ) {
```
Comment (samwilson) [inline src/WikisourceApi.php]: Rather than handling this here, you could edit the Sparql (see my comment above).

File: src/WikisourceApi.php [hunk 4]
```diff
@@ -104,15 +104,24 @@ public function fetchWikisources( $cacheLifetime = null ) {
...
 				// Wikimedia language code.
 				. "?item wdt:P424 ?langCode . "
 				// Language of work or name.
 				. "?item wdt:P407 ?lang . "
 				// RDF label of the language, in the language.
-				. "?lang rdfs:label ?langName . FILTER(LANG(?langName) = ?langCode || ?langCode = 'mul') . " . "}";
+				. "?lang rdfs:label ?langName . FILTER(LANG(?langName) = ?langCode) ."
+				// perform a union of the results
+				. " } UNION { "
+				// Instance of Wikisource language edition.
```
Comment (samwilson) [inline src/WikisourceApi.php]: You could add `BIND( "Multilingual Wikisource" AS ?langName )` to hard-code the language name. That'd fix the warning below and make things slightly simpler.

File: src/WikisourceApi.php [hunk 5]
```diff
@@ -104,15 +104,24 @@ public function fetchWikisources( $cacheLifetime = null ) {
...
+				"SELECT ?item ?langCode ?langName WHERE { {"
 				// Instance of Wikisource language edition.
 				. "?item wdt:P31 wd:Q15156455 . "
 				// Wikimedia language code.
 				. "?item wdt:P424 ?langCode . "
 				// Language of work or name.
 				. "?item wdt:P407 ?lang . "
 				// RDF label of the language, in the language.
-				. "?lang rdfs:label ?langName . FILTER(LANG(?langName) = ?langCode || ?langCode = 'mul') . " . "}";
+				. "?lang rdfs:label ?langName . FILTER(LANG(?langName) = ?langCode) ."
```
Comment (samwilson) [inline src/WikisourceApi.php]: An alternative way to get the extra result would be to add a separate clause here, e.g.:

```sparql
  FILTER(LANG(?langName) = ?langCode || (?langCode = 'mul' && LANG(?langName) = 'en') )
```

**Pull request merged**